src/devices_models/devices/common/activepower_constraints.jl

```diff
@@ -0,0 +1,116 @@
+
+struct ActivePowerConstraintsInputs
+    limits::Function
+    range_constraint::Function
+    multiplier::Function
+    timeseries_func::Function
+    timeseries_func_kwargs::Dict{Symbol, Any}
+    parameter_name::Union{Nothing, String}
+    constraint_name::String
+    variable_name::String
+    bin_variable_name::Union{Nothing, String}
+    forecast_label::String
+
+    # Force using kwargs to avoid ordering mistakes.
+    function ActivePowerConstraintsInputs(;
+        limits,
+        range_constraint,
+        multiplier,
+        timeseries_func,
+        timeseries_func_kwargs = Dict{Symbol, Any}(),
+        parameter_name,
+        constraint_name,
+        variable_name,
+        bin_variable_name,
+        forecast_label,
+    )
+        return new(
+            limits,
+            range_constraint,
+            multiplier,
+            timeseries_func,
+            timeseries_func_kwargs,
+            parameter_name,
+            constraint_name,
+            variable_name,
+            bin_variable_name,
+            forecast_label,
+        )
+    end
+end
+
+function ActivePowerConstraintsInputs(
+    ::Type{T},
+    ::Type{U},
+    use_parameters::Bool,
+    use_forecasts::Bool,
+) where {T <: PSY.Device, U <: AbstractDeviceFormulation}
+    error("ActivePowerConstraintsInputs is not implemented for type $T/$U")
+end
+
+"""
+Default implementation to add activepower constraints.
+
+Users of this function must implement a method for [`ActivePowerConstraintsInputs`](@ref).
+Users may also implement custom activepower_constraints! methods.
+"""
+function activepower_constraints!(
+    psi_container::PSIContainer,
+    devices::IS.FlattenIteratorWrapper{T},
+    model::DeviceModel{T, U},
+    ::Type{<:PM.AbstractPowerModel},
+    feedforward::Union{Nothing, AbstractAffectFeedForward},
+) where {T <: PSY.Device, U <: AbstractDeviceFormulation}
+    use_parameters = model_has_parameters(psi_container)
+    use_forecasts = model_uses_forecasts(psi_container)
+    inputs = ActivePowerConstraintsInputs(T, U, use_parameters, use_forecasts)
+
+    # TODO
+    # Tests for RenewableGen violate this.
+    #if use_parameters && !use_forecasts
+    #    error("use_parameters && !use_forecasts is not supported")
+    #end
```

Comment:
Please advise here.

Reply:
Let's add an argument error in the creation of psi_container when these 2 kwargs are introduced